src/textual/parser.rs

```diff
@@ -91,25 +91,26 @@ impl ParseErrorKind {
 
 /// A parser for terms using the Prolog-like syntax of the
 /// [TextualUniverse](super::TextualUniverse).
-pub struct Parser<'u> {
-    symbols: &'u mut SymbolStore,
+pub struct Parser<T: SymbolStorage> {
+    symbols: T,
 }
 
-impl<'a> Parser<'a> {
-    pub fn new(symbols: &'a mut SymbolStore) -> Self {
-        Self { symbols }
-    }
-
-    // //////////////////////////////// PUBLIC PARSER ////////////////////////////////
-
-    pub fn parse_query_str(&mut self, query: &str) -> Result<Query, ParseError> {
+impl<'a> Parser<SymbolOverlay<'a>> {
+    pub fn parse_query_str(mut self, query: &str) -> Result<UniverseQuery<'a>, ParseError> {
         let mut tokens = TokenStream::new(query);
         let mut scope = VarScope::new();
         let goals = self.parse_conjunction1(&mut tokens, &mut scope)?;
         self.expect_eof(&mut tokens)?;
-        Ok(Query::new(goals, Some(scope)))
+        Ok(UniverseQuery::new(self.symbols, Query::new(goals, Some(scope))))
+    }
+}
```

Comment:
It's a bit odd that the parser needs to know about the concrete overlay here.

Since these symbols are mainly relevant when using the `TextualUniverse`, I'd still make this return `Query`, and only build the `UniverseQuery` back in the `TextualUniverse`. Then only `prepare_query` needs to know about these overlays.

Reply:
I changed that, but I don't really like the outcome. The sets of symbols are in reality tightly bound to the parsed query, and this introduces a place where they may get mixed up.
I guess I don't see a reason why the Parser is a type on its own when it doesn't really carry over any state between consecutive parses. I'd make toplevel/public calls like parse_query_str consume it, so that another parse call can't reuse the same symbols store.

Reply:
I can see two use cases that I think should be supported:
1. Yours, where you want to have a `&TextualUniverse` that can be shared across many queries concurrently.
2. Being able to relate the results of multiple separate queries to each other. E.g. if I have a first query that mentions `foo`, and get some result, then I would like the second query that also mentions `foo` to produce the same result.

The second one is not possible with this design where you force query parsing to always use an overlay that's then consumed in the process.

---

With this new `SymbolStorage` trait, we basically have an abstraction for string interning, and I don't think the parser should force one particular way of how the consumer of the API should intern their strings.

In this setup, the consumer of this API is responsible for maintaining the relationship between the parsed structures (not just queries, but also rules), and the symbol storage.

The higher-level API, that maintains this mapping for the user of the interface would be the `TextualUniverse` type then, which is why I think that's where the `UniverseQuery` should be produced as well, rather than directly in the parser.